crypto/bio/bio.c

```diff
@@ -68,14 +68,59 @@
 
 #include "../internal.h"
 
-#define HAS_CALLBACK(b) ((b)->callback_ex != NULL)
+#define HAS_CALLBACK(b) ((b)->callback_ex != NULL || (b)->callback != NULL)
+
+//   |callback_fn_wrap_ex| adapts the legacy callback interface |BIO_callback_fn| to the
+//   extended callback interface |BIO_callback_fn_ex|. This function should only be
+//   called when |callback_ex| is not available and the legacy callback is set.
+//
+//   |len| and |processed| parameters from the extended interface are ignored since
+//   they are not supported by the legacy callback interface.
+//
+//   Returns -1 on NULL |BIO| or callback, otherwise returns the result of the legacy
+//   callback.
+static long callback_fn_wrap_ex(BIO *bio, int oper, const char *argp,
+                              size_t len, int argi, long argl, int ret,
+                              size_t *processed) {
+  if (bio == NULL || bio->callback == NULL) {
+    return -1;
+  }
+  return bio->callback(bio, oper, argp, argi, argl, ret);
+}
+
+//   |get_callback| returns the appropriate callback function for a given |BIO|, preferring
+//   the extended interface |callback_ex| over the legacy interface. This function
+//   should only be called after verifying a callback exists via |HAS_CALLBACK|.
+//
+//   When only the legacy callback is available, it is wrapped in the extended format
+//   via |callback_fn_wrap_ex| to provide a consistent interface. The extended callback
+//   provides additional parameters for length and bytes processed tracking.
+//
+//   Returns the |callback_ex| function if available, a wrapped legacy callback if only
+//   |callback| is set, or NULL if no callbacks are set (should not occur when called
+//   after |HAS_CALLBACK|).
+static BIO_callback_fn_ex get_callback(BIO *bio) {
+  if (bio->callback_ex != NULL) {
+    return bio->callback_ex;
+  }
+  if (bio->callback != NULL) {
+    // Wrap old-style callback in extended format
+    return callback_fn_wrap_ex;
+  }
+  return NULL;  // Explicit return, though this case should never happen if called after HAS_CALLBACK
+}
```

Comment:
Since this should never happen, we should add an `assert` so that if it occurs during testing, the tests will fail:
```
assert(false);
```

Reply:
It might be better to just not use the macro and just use the return value of `get_callback` to determine what to do.